Gordon360/Services/ProfileService.cs

```diff
@@ -267,6 +278,114 @@ public async Task UpdateCustomProfileAsync(string username, string type, CUSTOM_
+
+        // Privacy settings are generally heirarchical from bypassing all
+        // privacy settings to honoring all privacy settings:
+        //   SiteAdmin & Police -> FacStaff -> Students -> Alumni
+
+        // Find the account belonging to person whose profile we are accessing and get their
+        // privacy settings
+        var account = accountService.GetAccountByUsername(restricted_profile.AD_Username);
+        var privacy = context.UserPrivacy_Settings.Where(up_s => up_s.gordon_id == account.GordonID);
+
+        // Determing the viewer and profile user types
+        bool viewerIsSiteAdmin = viewerGroups.Contains(AuthGroup.SiteAdmin);
+        bool viewerIsPolice = viewerGroups.Contains(AuthGroup.Police);
+        bool viewerIsFacStaff = viewerGroups.Contains(AuthGroup.FacStaff);
+        bool viewerIsStudent = viewerGroups.Contains(AuthGroup.Student);
+        bool viewerIsAlumni = viewerGroups.Contains(AuthGroup.Alumni);
+        bool profileIsFacStaff = restricted_profile.PersonType.Contains(FACSTAFF_PROFILE);
+        bool profileIsStudent = restricted_profile.PersonType.Contains(STUDENT_PROFILE);
+        bool profileIsAlumni = restricted_profile.PersonType.Contains(ALUMNI_PROFILE);
+
+        foreach (UserPrivacy_Settings row in privacy)
+        {
+            if ((viewerIsSiteAdmin || viewerIsPolice) && row.Visibility != "Public")
+            {
+                MarkAsPrivate(restricted_profile, row.Field);
+            }
+            else if (viewerIsFacStaff)
+            {
+                if (profileIsFacStaff && row.Visibility == "Private")
+                {
+                    MakePrivate(restricted_profile, row.Field);
+                }
+                else if ((profileIsStudent || profileIsAlumni) && row.Visibility != "Public")
+                {
+                    MarkAsPrivate(restricted_profile, row.Field);
+                }
+            }
+            else if (viewerIsStudent && row.Visibility != "Public")
+            {
+                MakePrivate(restricted_profile, row.Field);
+            }
+            else if (viewerIsAlumni && row.Visibility != "Public")
+            {
+                MakePrivate(restricted_profile, row.Field);
+            }
+        }
+
+        return restricted_profile;
+    }
+
+    /// <summary>
+    /// get privacy setting for particular user
+    /// </summary>
+    /// <param name="username">AD username</param>
+    /// <returns>all privacy setting of the particular user</returns>
+    public IEnumerable<UserPrivacyViewModel> GetPrivacySettingAsync(string username)
+    {
+        var account = accountService.GetAccountByUsername(username);
+
+        // select all privacy settings
+        var privacy = context.UserPrivacy_Settings.Where(up_s => up_s.gordon_id == account.GordonID).Select(up_s => (UserPrivacyViewModel)up_s);
+
+        return privacy;
+    }
+
+    /// <summary>
+    /// privacy setting of some piece of personal data for user.
```

Comment:
Let's make this clearer.  Perhaps "Set the user's privacy preference for a piece of data"?

Reply:
Actually, it appears to be a list of pieces of data.

Reply:
Comments adjusted.